### PR TITLE
Add Map support to Kotlin generator

### PR DESCRIPTION
**Pull request by Egorand**
Part of #728 

Missing support for `RuntimeMessageAdapter`, which is not compatible with Kotlin protos as is.

### CODE REVIEW

File: wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt [hunk 1]
```diff
@@ -286,23 +286,27 @@ class KotlinGenerator private constructor(
...
-      var fieldClass: TypeName = typeName(field.type())
+      var fieldClass: TypeName
       val fieldName = nameAllocator.get(field)
       var defaultValue = CodeBlock.of("null")
 
       when {
-        field.isOptional -> {
-          fieldClass = fieldClass.asNullable()
-        }
+        field.isOptional -> fieldClass = typeName(field.type()).asNullable()
```
Comment (mc-lovin) [inline wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt]: nit can declare a variable for `field.type()`

File: wire-tests/src/test/proto-kotlin/com/squareup/wire/protos/kotlin/map/Mappy.kt [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+            override fun encodedSize(value: Mappy): Int = thingsAdapter.encodedSizeWithTag(1, value.things) +
+                value.unknownFields.size
+
+            override fun encode(writer: ProtoWriter, value: Mappy) {
+                thingsAdapter.encodeWithTag(writer, 1, value.things)
+                writer.writeBytes(value.unknownFields)
+            }
+
+            override fun decode(reader: ProtoReader): Mappy {
+                var things = mutableMapOf<String, Thing>()
```
Comment (mc-lovin) [inline wire-tests/src/test/proto-kotlin/com/squareup/wire/protos/kotlin/map/Mappy.kt]: should this be val?

Comment (mc-lovin) [reply wire-tests/src/test/proto-kotlin/com/squareup/wire/protos/kotlin/map/Mappy.kt]: if yes should be changed for list declarations as well, can be done in a separate pr

File: wire-tests/src/test/java/com/squareup/wire/KotlinMapTest.kt [hunk 3]
```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2016 Square Inc.
```
Comment (mc-lovin) [inline wire-tests/src/test/java/com/squareup/wire/KotlinMapTest.kt]: 2018

File: wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt [hunk 4]
```diff
@@ -362,13 +366,28 @@ class KotlinGenerator private constructor(
         .addFunction(encodedSizeFun(type))
         .addFunction(encodeFun(type))
         .addFunction(decodeFun(type))
-        .build()
+
+    for (field in type.fields()) {
+      if (field.type().isMap) {
+        val adapterType = ProtoAdapter::class.asTypeName().parameterizedBy(Map::class.asTypeName()
+            .parameterizedBy(typeName(field.type().keyType()), typeName(field.type().valueType())))
```
Comment (mc-lovin) [inline wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt]: nit can extract `keyType` and `valueType` into variables

File: wire-tests/src/test/proto-kotlin/com/squareup/wire/protos/kotlin/map/Mappy.kt (hunk 2 above)
Comment (Egorand) [reply wire-tests/src/test/proto-kotlin/com/squareup/wire/protos/kotlin/map/Mappy.kt]: Yes, I'll open an issue

**Pull request merged**